// moq.ts/esm2015/internal/lib/interactions.js
/*eslint-disable max-classes-per-file*/
export class Interaction {
    constructor(name, args) {
        this.name = name;
        this.args = args;
    }
}
/**
 * This class represents an invocation of a named function.
 * It provides access to the name of function and list of parameters.
 */
export class NamedMethodInteraction extends Interaction {
    constructor(name, args) {
        super(name, args);
    }
}
/**
 * This class represents an invocation of a function.
 * It provides access to the list of parameters.
 */
export class MethodInteraction extends Interaction {
    constructor(args) {
        super(undefined, args);
    }
}
/**
 * This class represents a property accessing.
 * It provides access to the name of property.
 */
export class GetPropertyInteraction extends Interaction {
    constructor(name) {
        super(name, undefined);
    }
}
/**
 * This class represents a property write interaction.
 * It provides access to the name of property and the value.
 */
export class SetPropertyInteraction extends Interaction {
    constructor(name, value) {
        super(name, [value]);
        this.value = value;
    }
}
/**
 * This class represents applying of [in operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in).
 * It provides access to the name of property.
 */
export class InOperatorInteraction extends Interaction {
    constructor(name) {
        super(name, undefined);
    }
}
