// moq.ts/esm2015/internal/lib/expression-matchers/expression.matcher.js
import { GetPropertyInteraction, InOperatorInteraction, MethodInteraction, NamedMethodInteraction, SetPropertyInteraction } from "../interactions";
import { ExpectedGetPropertyExpression, ExpectedInOperatorExpression, ExpectedMethodExpression, ExpectedNamedMethodExpression, ExpectedSetPropertyExpression } from "../expected-expressions/expected-expressions";
import { It } from "../expected-expressions/expression-predicates";
/**
 * @hidden
 */
export class ExpressionMatcher {
    constructor(getPropertyExpressionMatcher, setPropertyExpressionMatcher, methodExpressionMatcher, namedMethodExpressionMatcher, inOperatorExpressionMatcher) {
        this.getPropertyExpressionMatcher = getPropertyExpressionMatcher;
        this.setPropertyExpressionMatcher = setPropertyExpressionMatcher;
        this.methodExpressionMatcher = methodExpressionMatcher;
        this.namedMethodExpressionMatcher = namedMethodExpressionMatcher;
        this.inOperatorExpressionMatcher = inOperatorExpressionMatcher;
    }
    matched(left, right) {
        if (left === right)
            return true;
        if (right === undefined)
            return true;
        if (left instanceof GetPropertyInteraction && (right instanceof ExpectedGetPropertyExpression || right instanceof It)) {
            return this.getPropertyExpressionMatcher.matched(left, right);
        }
        if (left instanceof SetPropertyInteraction && (right instanceof ExpectedSetPropertyExpression || right instanceof It)) {
            return this.setPropertyExpressionMatcher.matched(left, right);
        }
        if (left instanceof InOperatorInteraction && (right instanceof ExpectedInOperatorExpression || right instanceof It)) {
            return this.inOperatorExpressionMatcher.matched(left, right);
        }
        if (left instanceof MethodInteraction && (right instanceof ExpectedMethodExpression || right instanceof It)) {
            return this.methodExpressionMatcher.matched(left, right);
        }
        if (left instanceof NamedMethodInteraction && (right instanceof ExpectedNamedMethodExpression || right instanceof It)) {
            return this.namedMethodExpressionMatcher.matched(left, right);
        }
        return false;
    }
}
