// moq.ts/esm2015/internal/lib/presets/preset-builder.js
import { MimicsPreset } from "./presets/mimics.preset";
import { ReturnsPreset } from "./presets/returns.preset";
import { ThrowsPreset } from "./presets/throws.preset";
import { CallbacksPreset } from "./presets/callbacks.preset";
import { PlayTimes } from "../playables/play-times";
/**
 * The default implementation of {@link IPresetBuilder} interface.
 * Is it not intended to be used outside of the moq library.
 *
 * @hidden
 */
export class PresetBuilder {
    constructor(mock, set, target, playable = PlayTimes.Always()) {
        this.mock = mock;
        this.set = set;
        this.target = target;
        this.playable = playable;
    }
    mimics(origin) {
        const preset = new MimicsPreset(this.playable, this.target, origin);
        this.set(preset);
        return this.mock;
    }
    returns(value) {
        const preset = new ReturnsPreset(this.playable, this.target, value);
        this.set(preset);
        return this.mock;
    }
    throws(exception) {
        const preset = new ThrowsPreset(this.playable, this.target, exception);
        this.set(preset);
        return this.mock;
    }
    callback(callback) {
        const preset = new CallbacksPreset(this.playable, this.target, callback);
        this.set(preset);
        return this.mock;
    }
    play(playable) {
        this.playable = playable;
        return this;
    }
}
