// moq.ts/esm2015/internal/lib/equal-matchers/equal.matcher.js
/**
 * @hidden
 */
export class EqualMatcher {
    constructor(typesMatcher, commonTypeProvider, primitiveMatcher, objectMatcher, functionMatcher) {
        this.typesMatcher = typesMatcher;
        this.commonTypeProvider = commonTypeProvider;
        this.primitiveMatcher = primitiveMatcher;
        this.objectMatcher = objectMatcher;
        this.functionMatcher = functionMatcher;
    }
    matched(left, right) {
        if (this.typesMatcher.matched(left, right) === false) {
            return false;
        }
        switch (this.commonTypeProvider.ofType(left, right)) {
            case "undefined":
                return this.primitiveMatcher.matched(left, right);
            case "object":
                return this.objectMatcher.matched(left, right);
            case "boolean":
                return this.primitiveMatcher.matched(left, right);
            case "number":
                return this.primitiveMatcher.matched(left, right);
            case "string":
                return this.primitiveMatcher.matched(left, right);
            case "function":
                return this.functionMatcher.matched(left, right);
            case "symbol":
                return this.primitiveMatcher.matched(left, right);
            case "bigint":
                return this.primitiveMatcher.matched(left, right);
        }
    }
}
