// moq.ts/esm2015/public_api.js
/*
 * Public API Surface of moq.ts
 */
export * from "./lib/moq";
export * from "./lib/mock";
export * from "./lib/dump";
export * from "./lib/interactions";
export * from "./lib/times";
export * from "./lib/playables/play-times";
export * from "./lib/expected-expressions/expected-expressions";
export * from "./lib/expected-expressions/expression-predicates";
export * from "./lib/playables/play-times";
export * from "./lib/playables/playable-exactly";
export * from "./lib/playables/playable-once";
export * from "./lib/playables/playable-never";
export * from "./lib/playables/playable-always";
export * from "./lib/playables/playable-sequence";
export * from "./lib/injector/default-injector.config";
export * from "./lib/injector/equal-matching-injector.config";
export * from "./lib/injector/injector.factory";
export * from "./lib/injector/moq.injection-token";
export * from "./lib/mock-options/mock-options.injection-token";
export * from "./lib/equal-matchers/object-matchers.injection-token";
export * from "./lib/equal-matchers/object-matcher.type";
export * from "./lib/static.injector/metadata";
export * from "./lib/static.injector/interface/injector";
export { forwardRef, isForwardRef } from "./lib/static.injector/forward_ref";
export { Injector } from "./lib/static.injector/injector";
export { INJECTOR, NullInjector } from "./lib/static.injector/injector_compatibility";
export { getClosureSafeProperty } from "./lib/static.injector/property";
export * from "./lib/static.injector/injection_token";
