// moq.ts/esm2015/lib/formatters/expression-formatter.js
import { GetPropertyInteraction, InOperatorInteraction, MethodInteraction, NamedMethodInteraction, SetPropertyInteraction } from "../interactions";
import { It } from "../expected-expressions/expression-predicates";
/**
 * @hidden
 */
export class ExpressionFormatter {
    constructor(getPropertyFormatter, setPropertyFormatter, methodFormatter, namedMethodFormatter, constantFormatter, inOperatorFormatter) {
        this.getPropertyFormatter = getPropertyFormatter;
        this.setPropertyFormatter = setPropertyFormatter;
        this.methodFormatter = methodFormatter;
        this.namedMethodFormatter = namedMethodFormatter;
        this.constantFormatter = constantFormatter;
        this.inOperatorFormatter = inOperatorFormatter;
    }
    format(expression) {
        if (expression instanceof GetPropertyInteraction)
            return this.getPropertyFormatter.format(expression);
        if (expression instanceof SetPropertyInteraction)
            return this.setPropertyFormatter.format(expression);
        if (expression instanceof InOperatorInteraction)
            return this.inOperatorFormatter.format(expression);
        if (expression instanceof MethodInteraction)
            return this.methodFormatter.format(expression);
        if (expression instanceof NamedMethodInteraction)
            return this.namedMethodFormatter.format(expression);
        if (expression instanceof It)
            return this.constantFormatter.format(expression);
        return undefined;
    }
}
