// moq.ts/esm2015/internal/lib/expected-expressions/expected-expression-reflector.js
import { It } from "./expression-predicates";
import { ExpectedGetPropertyExpression, ExpectedInOperatorExpression, ExpectedMethodExpression, ExpectedNamedMethodExpression, ExpectedSetPropertyExpression } from "./expected-expressions";
/**
 * This class reflects an expression to an expression tree representation.
 *
 * @example
 * ```typescript
 *
 * const arg = 'argument';
 * const reflector = new ExpectedExpressionReflector();
 * const actual = reflector.reflect<any>(instance => instance(arg));
 *
 * const expected = new ExpectedMethodExpression([arg]);
 * expect(actual).toEqual(expected);
 * ```
 *
 * For more examples check
 * [unit tests for this class]
 * (https://github.com/dvabuzyarov/moq.ts/blob/master/tests.unit/expected-expressions/expected-expression-reflector.UnitTests.ts)
 */
export class ExpectedExpressionReflector {
    /**
     * Reflects the provided code as an expression tree.
     */
    reflect(expression) {
        this.reflectedInfo = undefined;
        const proxy = this.expressionProxy();
        const predicate = expression(proxy);
        return predicate instanceof It && (this.reflectedInfo instanceof ExpectedSetPropertyExpression) === false
            ? predicate : this.reflectedInfo;
    }
    expressionProxy() {
        const options = {
            get: (target, name) => {
                this.reflectedInfo = new ExpectedGetPropertyExpression(name);
                return (...args) => {
                    this.reflectedInfo = new ExpectedNamedMethodExpression(name, args);
                };
            },
            set: (target, name, value) => {
                this.reflectedInfo = new ExpectedSetPropertyExpression(name, value);
                return true;
            },
            apply: (target, thisArg, args) => {
                this.reflectedInfo = new ExpectedMethodExpression(args);
            },
            has: (target, name) => {
                this.reflectedInfo = new ExpectedInOperatorExpression(name);
                return true;
            }
        };
        return new Proxy(() => undefined, options);
    }
}
